### DISCUSSION TITLE
Attestation Certificate Validation at Registration Ceremony

### DISCUSSION THREAD

**Discussion by ThaminduDilshan** [webauthn4j/webauthn4j#635] [Q&A] [answered]
Hi,

**TLDR;;**
Is it possible to perform attestation certificate validations using webauthn manager `validate` function? If so, please provide some context on the implementation. I'm following the implementation in [this doc](https://webauthn4j.github.io/webauthn4j/en/#webauthn-attestation-verification) and observed that some FIDO conformance test cases related to incorrect certificate data are failing.

**Question**
I'm interested in performing attestation validations when registering FIDO devices using this library. I did follow the sample code block from [webauthn4j docs](https://webauthn4j.github.io/webauthn4j/en/#webauthn-attestation-verification) and implemented the server code.

These scenarios are tested using the FIDO conformance test tool and observed that some test cases related to incorrect certificate are failing (F1, F2 and F4 in `Server-ServerAuthenticatorAttestationResponse-Resp-9 Test server processing "tpm" attestation`). These scenarios are validated using the webauthn4j attestation validation step and hasn't thrown any errors.

I'm unclear here why the certificate validations are not performed. Was wondering whether there are some missing configs or data in the server implementation when performing `webAuthnManager.validate`?
I followed all the configs mentioned in the referenced doc.

Cheers,
Thamindu

**Comment by ynojima** [comment, marked as answer]
How do you initialize `WebAuthnManager` instance?
To validate attestation statements, you need to provide attestationStatementValidator(s) like this way:
```
new WebAuthnManager(
                Arrays.asList(
                        new PackedAttestationStatementValidator(),
                        new FIDOU2FAttestationStatementValidator(),
                        new AndroidKeyAttestationStatementValidator(),
                        new AndroidSafetyNetAttestationStatementValidator(),
                        new TPMAttestationStatementValidator(),
                        new AppleAnonymousAttestationStatementValidator(),
                        new NoneAttestationStatementValidator()
                ),
                certPathTrustworthinessValidator,
                new DefaultSelfAttestationTrustworthinessValidator(),
                objectConverter
        );
```
Related doc section:
https://webauthn4j.github.io/webauthn4j/en/#attestation-statement-validation

I have FIDO Server sample that passes TPM test cases except P-1 and P-3. Please check this code for reference, and ask me anything.
https://github.com/webauthn4j/webauthn4j-spring-security/tree/master/samples/fido-server-conformance-test-app

Note: Regarding P-1 and P-2, there is a known issue on FIDO Conformance tools end: https://github.com/fido-alliance/conformance-test-tools-resources/issues/618

**Comment by ThaminduDilshan** [comment]
I have used `WebAuthnManager.createNonStrictWebAuthnManager()` function to initiate the webauthn manager object. With the provided context, I guess my implementation is missing these required attestation validators.

One point I'm still not clear is how to construct the `certPathTrustworthinessValidator` parameter. By looking at the framework implementation, I found several concrete implementations of  `CertPathTrustworthinessValidator` interface and I guess I can use `TrustAnchorCertPathTrustworthinessValidator` object for my implementation. Can you provide some information on how to create this object?

**Reply by ynojima** [reply]
> I have used WebAuthnManager.createNonStrictWebAuthnManager() function to initiate the webauthn manager object. With the provided context, I guess my implementation is missing these required attestation validators.

Yes, `WebAuthnManager` created from `.createNonStrictWebAuthnManager` skip attestation statements validation.

`TrustAnchorCertPathTrustworthinessValidator` id deprecated from `0.19.0.RELEASE`.
It is better to uses `DefaultCertPathTrustworthinessValidator` instead.

Here is a sample how to use `DefaultCertPathTrustworthinessValidator` :
https://github.com/webauthn4j/webauthn4j-spring-security/blob/master/samples/fido-server-conformance-test-app/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/app/config/WebSecurityBeanConfig.java#L185-L191

**Reply by ThaminduDilshan** [reply]
Thanks for the information. I was able to implement this as mentioned and pass the attestation certificate validations.
